### PR TITLE
support more attributes from the Encoding structure

### PR DESCRIPTION
**Pull request by clems4ever**
MiniLM requires the attention mask to perform the mean pooling operation as can be seen at
https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2

### CODE REVIEW

File: tokenizers.h [hunk 1]
```diff
@@ -3,7 +3,10 @@
 
 struct Buffer {
   uint32_t *ids;
+  uint32_t *type_ids;
   char *tokens;
+  uint32_t *special_tokens_mask;
+  uint32_t *attention_mask;
```
Comment (daulet) [inline tokenizers.h:9]: I kind of regret adding this struct to expose `char *tokens`. It made encode call about ~2x slower due to extra allocations and most users don't care about string tokens. For your feature, can you make it a distinct rust/c method for encode? The go interface could still be the same (perhaps with additional flags) to request one implementation over the other, however that way users that don't use these features won't be slowed down by this extra allocations.

Comment (clems4ever) [reply tokenizers.h:9]: I agree, when writing this code I wondered whether we should have some way of parameterizing what gets back to the Go API so that you only pay the price for exactly what you need. I propose to introduce a new EncodeWithOptions function to which we can pass the list of attributes (as options) in order parameterize which attributes are to get retrieved. Would that be ok?

Comment (clems4ever) [reply tokenizers.h:9]: Obviously we would keep the Encode function as it was before this change to not impact its performance.

Comment (riccardopinosio) [reply tokenizers.h:9]: Hello, I think this makes a lot of sense. For instance, in our case we needed not only the attention mask, but also the token_type_ids to be able to use an onnx version of the MiniLM model. So we essentially forked the repo and just added rust code to the encode method to return those and the attention mask. It would be great to have an EncodeWIthOptions to customize the output

Comment (riccardopinosio) [reply tokenizers.h:9]: A similar thing would be handy when reading the tokenizer in, e.g., to set truncation and padding. Maybe the from_file and from_bytes methods could support an option struct?

Comment (clems4ever) [reply tokenizers.h:9]: I'll update the PR with the EncodeWithOptions very soon.

Regarding the configuration of the tokenizer, may I suggest you open a new dedicated issue to discuss that topic specifically?

Comment (clems4ever) [reply tokenizers.h]: @daulet , @riccardopinosio , done. Let me know what you think.

Comment (riccardopinosio) [reply tokenizers.h]: @clems4ever looks great to me.
@daulet can we merge this one?

Comment (daulet) [reply tokenizers.h]: Sorry this was buried in my github notifications.

I like how you were able to reuse the same method while avoiding allocations. Do you mind resolving conflicts and rebasing? I added CI so we can quickly run the tests.

File: tokenizer.go [hunk 2]
```diff
@@ -54,27 +54,139 @@ func (t *Tokenizer) Close() error {
...
+	slice := make([]uint32, len)
+	for i, v := range arr {
+		slice[i] = uint32(v)
+	}
+	return slice
+}
+
+func CBool(b bool) C.bool {
+	return C.bool(b)
+}
```
Comment (daulet) [inline tokenizer.go:87]: i feel this doesn't deserve to be a func

File: tokenizer_test.go [hunk 3]
```diff
@@ -227,8 +364,8 @@ func BenchmarkEncodeNTimes(b *testing.B) {
 	expected := []uint32{2829, 4419, 14523, 2058, 1996, 13971, 3899}
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		tokens, _ := tk.Encode("brown fox jumps over the lazy dog", false)
-		assert.Equal(b, expected, tokens)
+		ids, _ := tk.Encode("brown fox jumps over the lazy dog", false)
```
Comment (daulet) [inline tokenizer_test.go:367]: do you mind running before and after benchmarks?

File: tokenizers.h (hunk 1 above)
Comment (RJKeevil) [reply tokenizers.h]: @clems4ever just making sure you saw this, would love to have this in the main repo, then I can kill off my fork

File: tokenizer_test.go (hunk 3 above)
Comment (clems4ever) [reply tokenizer_test.go:367]: Sure. I will post the results in the PR.

File: tokenizer.go (hunk 2 above)
Comment (clems4ever) [reply tokenizer.go:87]: ok, I will remove it.

**Comment by clems4ever** [comment]
Before

$ go test . -bench=. -benchmem -benchtime=10s
goos: linux
goarch: amd64
pkg: github.com/daulet/tokenizers
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkEncodeNTimes-8           885213             13053 ns/op             224 B/op         11 allocs/op
BenchmarkEncodeNChars-8         1000000000               2.351 ns/op           0 B/op          0 allocs/op
BenchmarkDecodeNTimes-8          2108638              5758 ns/op              96 B/op          3 allocs/op
BenchmarkDecodeNTokens-8        15591064               761.3 ns/op             7 B/op          0 allocs/op
PASS
ok      github.com/daulet/tokenizers    59.096s


After

$ go test . -bench=. -benchmem -benchtime=10s
goos: linux
goarch: amd64
pkg: github.com/daulet/tokenizers
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkEncodeNTimes-8           935011             12774 ns/op             232 B/op         12 allocs/op
BenchmarkEncodeNChars-8         1000000000               1.962 ns/op           0 B/op          0 allocs/op
BenchmarkDecodeNTimes-8          2098053              5676 ns/op              96 B/op          3 allocs/op
BenchmarkDecodeNTokens-8        15740354               742.0 ns/op             7 B/op          0 allocs/op
PASS
ok      github.com/daulet/tokenizers    57.765s

File: tokenizer.go (hunk 2 above)
Comment (clems4ever) [reply tokenizer.go]: done

File: tokenizer_test.go (hunk 3 above)
Comment (clems4ever) [reply tokenizer_test.go:386]: done

File: tokenizers.h (hunk 1 above)
Comment (clems4ever) [reply tokenizers.h]: PR rebased and benchmark results posted in the PR

**Review by daulet — APPROVED** [review]
Thank you for your contribution!
